prov/efa: Allow fork support and MR cache

rwespetal opened the pull request:
This series allows applications to use fork again by disabling the internal use of huge pages when fork support is requested and the registration cache is enabled. Applications that require the registration of regions backed by huge pages, fork support, and the registration cache are still not supported and will cause the domain open to fail. That mode will be supported in the future when changes to the linux kernel makes rdma-core fork support irrelevant in general by copying pages into the fork.

For convenience a new environment variable, `FI_EFA_FORK_SAFE`, was also added here and will call `ibv_fork_init()` when set.

Test description: Set the new environment variable and the rdma-core environment variables and verified that huge pages are not used when fork support is not enabled. Also verified that the check still works when another library calls `ibv_fork_init()` directly. Verified setting `RDMAV_HUGEPAGES_SAFE` returns an error from domain open. Finally, I added a fork to MPI ring_c; the handler is triggered when fork support is disabled, and the test passes when fork support is enabled.

wzamazon commented on prov/efa/src/rxr/rxr_init.c:
```diff
@@ -125,6 +129,9 @@ static void rxr_init_env(void)
 			    &rxr_env.efa_read_segment_size);
+	fi_param_get_bool(&rxr_prov, "fork_safe", &fork_safe);
+	if (fork_safe)
+		efa_fork_status = EFA_FORK_SUPPORT_ON;
```
If RDMAV_FORK_SAFE/IBV_FORK_SAFE is set, we also want to set fork support to be on

rwespetal replied:
Right, those do get checked in `efa_check_fork_enabled` but it'd be more intuitive to check it here. Let me move that check.

wzamazon replied:
Thanks!

wzamazon commented on prov/efa/src/efa_domain.c:
```diff
@@ -309,8 +324,23 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
+	 * buffers will not be used if fork support is on.
+	 */
+	if (efa_fork_status == EFA_FORK_SUPPORT_ON) {
+		ret = -ibv_fork_init();
```
I think call ibv_fork_init here is too late, in `efa_check_fork_enabled` we registered a pointer. If ibv_fork_init has not been called prior to that, this call would always fail.

rwespetal replied:
I'm not sure I understand the concern as `efa_check_fork_enabled` is called after this. But I don't see any harm in calling `ibv_fork_init` in `rxr_init_env` and can move it there.

wzamazon replied:
Sorry, I misread the change. (I did not notice that efa_fork_fork_enabled is to be moved).

But I would prefer ibv_fork_init to be called earlier. Thanks!

wzamazon reviewed (changes requested):
I think if fork_support is on, ibv_fork_init should be called earlier.

Also, I think we discussed that only register fork handler when fork_support is off.

rwespetal commented:
> Also, I think we discussed that only register fork handler when fork_support is off.

Sorry, I must have missed that. I'll make that change too.

wzamazon commented on prov/efa/src/efa_domain.c:
```diff
@@ -286,53 +286,38 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
-			ret = -FI_ENOMEM;
-			goto err_free_info;
-		}
+	if (app_mr_local || !efa_mr_cache_enable)
```
I think for normal code path to return early is a bit unusual. 

Would it make sense to put the whole section at the end into one function efa_mr_cache_init(), and do
```
if (!app_mr_local && efa_mr_cache_enable) {
     err = efa_mr_cache_init();
     if (err)
            efa_mr_cache_enable = false;
}

return 0;
```

rwespetal replied:
That is a much better idea, thanks for the suggestion.

wzamazon commented on prov/efa/src/efa_domain.c:
```diff
@@ -292,6 +309,9 @@ int efa_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
 	if (app_mr_local || !efa_mr_cache_enable)
 		return 0;
 
+	if (efa_check_fork_enabled(*domain_fid))
```
Maybe add `efa_fork_status == EFA_FORK_SUPPORT_OFF` to the condition?

If efa_fork_status is already ON, we do not want to run the check.
If efa_fork_status is UNNEEDED, we do not want to run the check either.

rwespetal commented:
I addressed the feedback and also added another commit to move the fork handler install to domain init so we can run the fork check before deciding to install it.

One thing to mention is moving the logic and call to ibv_fork_init to the provider init means that if 1) ibv_fork_init fails or 2) `RDMAV_HUGEPAGES_SAFE` and fork support is enabled we'll return a NULL provider (EFA is excluded) instead of failing the domain open. Let me know if that's not okay.

bwbarrett commented:
> One thing to mention is moving the logic and call to ibv_fork_init to the provider init means that if 1) ibv_fork_init fails or 2) `RDMAV_HUGEPAGES_SAFE` and fork support is enabled we'll return a NULL provider (EFA is excluded) instead of failing the domain open. Let me know if that's not okay.

This seems impossible for us to debug?  Particularly EFA not being selected because `RDMAV_HUGEPAGES_SAFE` is set to 1 while `FI_EFA_FORK_SAFE` is also set to 1.  We need to at least print a warning, if not abort.

rwespetal commented:
Yep, this will be a pain to debug. There is a warn but users will have to set `FI_LOG_LEVEL=warn` to see it. Perhaps the huge pages check should move back to the domain open. I'm less worried about ibv_fork_init failing but it is possible.

wzamazon commented:
> Perhaps the huge pages check should move back to the domain open. I'm less worried about ibv_fork_init failing but it is possible.

I did not think about this possibility when I suggested moving these to provider initialization. Considering these, I think moving both back to domain initialization is better.

rwespetal commented:
Sure, let me do that.

rwespetal commented:
Rebased and resolved a minor conflict, waiting on CI to merge.

rwespetal commented:
bot:aws:retest
